prov/usnic/src/usdf_dgram.c

```diff
@@ -531,17 +621,72 @@ usdf_dgram_prefix_sendv(struct fid_ep *fep, const struct iovec *iov, void **desc
 		send_iov[0].iov_base = ((char *) send_iov[0].iov_base +
 				USDF_HDR_BUF_ENTRY);
 		send_iov[0].iov_len -= USDF_HDR_BUF_ENTRY;
-		_usdf_dgram_send_iov_copy(ep, dest, send_iov, count, context);
+
+		return _usdf_dgram_send_iov_copy(ep, dest, send_iov, count,
+				context, ep->ep_tx_completion);
+	} else if (ep->e.dg.tx_op_flags & FI_INJECT) {
+		USDF_DBG_SYS(EP_DATA,
+				"given inject length (%zu) exceeds max inject length (%d)\n",
+				len, USD_SEND_MAX_COPY);
+		return -FI_ENOSPC;
 	}
 
-	return 0;
+	return _usdf_dgram_send_iov_prefix(ep, dest, iov, count, context,
+					   ep->ep_tx_completion);
 }
 
-ssize_t
-usdf_dgram_prefix_sendmsg(struct fid_ep *fep, const struct fi_msg *msg, uint64_t flags)
+ssize_t usdf_dgram_prefix_sendmsg(struct fid_ep *fep, const struct fi_msg *msg,
+		uint64_t flags)
 {
-	return usdf_dgram_prefix_sendv(fep, msg->msg_iov, msg->desc, msg->iov_count,
-				(fi_addr_t)msg->addr, msg->context);
+	struct iovec send_iov[USDF_DGRAM_MAX_SGE];
+	struct usd_dest *dest;
+	struct usdf_ep *ep;
+	uint8_t completion;
+	size_t len;
+
+	ep = ep_ftou(fep);
+	dest = (struct usd_dest *)(uintptr_t) msg->addr;
+	len = _usdf_iov_len(msg->msg_iov, msg->iov_count);
+	completion = ep->ep_tx_dflt_signal_comp || (flags & FI_COMPLETION);
+
+	/* both send paths require the use of a buffer the size of
+	 * USDF_DGRAM_MAX_SGE.
+	 */
+	if (msg->iov_count > USDF_DGRAM_MAX_SGE) {
+		USDF_DBG_SYS(EP_DATA, "max iov count exceeded: %zu\n",
+				msg->iov_count);
+		return -FI_ENOSPC;
+	}
+
+	if (len <= USD_SEND_MAX_COPY) {
+		/* _usdf_dgram_send_iov_copy isn't prefix aware and allocates
+		 * its own prefix. reorganize iov[0] base to point to data and
+		 * len to reflect data length.
+		 */
+		memcpy(send_iov, msg->msg_iov,
```

Comment:
Is this memcpy necessary? I don't think I should be modifying the input, but I could avoid doing a copy if so.